maven_iuvs/echelle.py

```diff
@@ -0,0 +1,1292 @@
+    """
+    coadded_lights = np.zeros_like(dark_sub[0, :, :]) # nan frames already controlled for.
+    total_frames = 0
+
+    for frame in range(dark_sub.shape[0]):
+        coadded_lights += dark_sub[frame]
+        total_frames += 1
+
+    return coadded_lights, total_frames
+
+
+def subtract_darks(light_fits, dark_fits, verbose=False):
+    """
+    Given matching light and dark fits, subtracts off the darks from lights.
+    """
+    first_dark, second_dark = get_dark_frames(dark_fits)
+    light_data = light_fits['Primary'].data
+
+    # Make the array to store dark-subtracted data
+    dark_subtracted = np.zeros_like(light_data)
+    
+    # Get rid of extra frames where light data are bad (broken or nan)
+
+    medians = []
+    good_frame_inds = [] # Frames which appear to have valid data
+    nan_light_inds = [] # frames with NaN
+    bad_light_inds = [] # Light frames which have some problem, i.e. oversaturation, but are not NaN
+    bad_dark_inds = [] # Possible to have a bad dark but a fine light, have to keep track of them separately
+
+    for i in range(0, light_data.shape[0]):
+        # reject light frames which are missing data (have NaN)
+        if np.isnan(light_data[i]).any():
+            # bad_frames += 1
+            nan_light_inds.append(i)
+            continue 
+
+        # reject frames where the median value is absurd - this indicates a broken frame
+        median_this_frame = np.median(light_data[i])
+
+        # We need to specially treat the possible case where the first frame could be broken. 
+        # Unlikely but possible. Currently done by comparing median with values known to be too high
+        # for typical detector image. TODO: Make this better and not rely on hard-coded value.
+        # Most medians are in the 100s due to the typical sky background values being similar.
+        if (not medians) & (median_this_frame >= 5000):
+            bad_light_inds.append(i)
+
+        # For all other light frames, we can compare to the stored median values of known good frames.
+        if (len(medians)>0) and (median_this_frame / np.median(medians) > 100): 
+            bad_light_inds.append(i)
+            continue
+
+        # At this point in the loop, the frame should have good data.
+        good_frame_inds.append(i)
+        medians.append(np.median(median_this_frame))
+
+    # Control for possibility of one dark frame or both containing NaN
+    if np.isnan(first_dark).any():
+        bad_dark_inds.append(0)
+
+    if np.isnan(second_dark).any():
+        bad_dark_inds.extend([i for i in range(1, light_data.shape[0]) if i not in bad_light_inds])
+
+    # collect all the indices of frames with some problem and remove an equivalent number of frames from the final matrix.
+    # Make sure it's a set (unique values only)
+    all_bad_inds = sorted(list(set([*nan_light_inds, *bad_light_inds, *bad_dark_inds])))
+    dark_subtracted = np.delete(dark_subtracted, all_bad_inds, axis=0)
```

Comment:
better to return NaN for invalid frames, that way return value always has same dimensions at input light array (rule of least surprise)

Reply:
Fixed by ae415e4a60f788a69c1d123945a1e143fd23e89a